package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,421 @@
+
+        # result
+        self.msds_by_particle = None
+        self.timeseries = None
+
+    def _prepare(self):
+        # self.n_frames only available here
+        # these need to be zeroed prior to each run() call
+        self.msds_by_particle = np.zeros((self.n_frames, self.n_particles))
+        self._position_array = np.zeros(
+            (self.n_frames, self.n_particles, self.dim_fac))
+        # self.timeseries not set here
+
+    def _parse_msd_type(self):
+        r""" Sets up the desired dimensionality of the MSD.
+
+        Parameters
+        ----------
+        self.msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Dimensions to be included in the MSD.
+
+        Returns
+        -------
+        self._dim : list
+            Array-like used to slice the positions to obtain desired
+            dimensionality
+        self.dim_fac : int
+            Dimension factor :math:`d` of the MSD.
+
+        """
+        keys = {'x': [0], 'y': [1], 'z': [2], 'xy': [0, 1],
+                'xz': [0, 2], 'yz': [1, 2], 'xyz': [0, 1, 2]}
+
+        self.msd_type = self.msd_type.lower()
+
+        try:
+            self._dim = keys[self.msd_type]
+        except KeyError:
+            raise ValueError(
+                'invalid msd_type: {} specified, please specify one of xyz, \
+                 xy, xz, yz, x, y, z'.format(self.msd_type))
+
+        self.dim_fac = len(self._dim)
+
+    def _single_frame(self):
+        r""" Constructs array of positions for MSD calculation.
+
+        Parameters
+        ----------
+        self.u : :class:`Universe`
+            MDAnalysis Universe
+
+        Returns
+        -------
+        self._position_array : :class:`numpy.ndarray`
+            Array of particle positions with respect to time
+            shape = (n_frames, n_particles, dim_fac)
+        """
+        # shape of position array set here, use span in last dimension
+        # from this point on
+        self._position_array[self._frame_index] = \
+            self._atoms.positions[:, self._dim]
+
+    def _conclude(self):
+        if self.fft:
+            self._conclude_fft()
+        else:
+            self._conclude_simple()
+
+    def _conclude_simple(self):
+        r""" Calculates the MSD via the simple "windowed" algorithm.
+
+        Parameters
+        ----------
+        self._position_array : :class:`numpy.ndarray`
+            Array of particle positions with respect to time
+            shape (n_frames, n_particles, dim_fac).
+
+        Returns
+        -------
+        self.timeseries : :class:`numpy.ndarray`
+            The MSD as a function of lag-time.
+
+        """
+        lagtimes = np.arange(1, self.n_frames)
+        for lag in lagtimes:
+            disp = self._position_array[:-lag, :, :] - \
+                self._position_array[lag:, :, :]
+            sqdist = np.square(disp, dtype=np.float64).sum(
+                axis=-1, dtype=np.float64)  # np.float64 required
```

Comment:
Is this second dtype required? I would have thought it would sum to the same dtype. If not, you could make this one line